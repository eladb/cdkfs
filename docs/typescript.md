# TypeScript Projects

To create a new TypeScript project, use `projen new typescript`:

```shell
npx projen new typescript
```

This will synthesize a standard project directory structure with some sample
code.

```shell
├── src
│   ├── index.ts
├── test
│   ├── hello.test.ts
├── node_modules
├── package.json
└── tsconfig.json
```

The default options will set up your `tsconfig` root directory as `src` and your
out directory as `lib`. By default the `test` directory is not compiled to
JavaScript 

The `projen new` command will also generate a `.projenrc.js` file which includes
the definition of your project with any options you specified in the command
line:

```js
const { typescript } = require('projen');
  const project = new typescript.TypeScriptProject({
    defaultReleaseBranch: 'main',
    name: 'my-project',
  });
project.synth();
```

> At this point, projenrc is in JavaScript, but in the future we plan to allow
> specifying your project definitions in TypeScript.

To modify your project definitions, edit `.projenrc.js` and run `projen` again
to re-synthesize your project. The following sections describe the various
features of your TypeScript project.

## Migrating your TypeScript Project to Projen

Projen provides some opinionated defaults that may be different than the defaults
you rely on in your existing TypeScript Project. The most important is the default
project structure. Projen expects all source code to be in the `src` directory and
will write compiled `.js` and `.d.ts` files to the `lib` directory. Test files are
expected in the `test` directory and not compiled to javascript. Instead, Projen
configures `Jest` to compile `.ts` files when testing.

As this is only one of many structures for valid TypeScript projects, your migrated
project may not work out-of-the-box with Projen. We recommend that you take the
time to fully migrate to the expected project structure so that you can take full
advantage of Projen management on your repository.

However, you can also configure your `.projenrc.js` file to reflect your current
project structure. A `TypeScriptProject` exposes all options in a `tsConfig` file
so you can manually set the structure you want. Here is an example for specifying a
structure where the TypeScript compiler finds all TypeScript files recursively in your
repository:

```js
const { typescript } = require('projen');
const project = new typescript.TypeScriptProject({
  ...
  tsConfig: {
    compilerOptions: {
      rootDir: '.',
      outDir: '.',
    },
    include: ['**/*.ts'], // all typescript files recursively
  }
});
```