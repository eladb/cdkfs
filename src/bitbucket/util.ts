

export function reduceRenderObject( obj:Record<string, any> ) {

  const result = Object.keys(obj).reduce( ( inp:Record<string, any>, key:string) => {

    if (obj[ key ] !== undefined) inp[ key ] = obj[ key ];

    return inp;
  }, {} );

  if ( Object.keys( result).length === 0 ) return undefined;

  return result;
}


export function reduceRenderArray( arr:any[]|undefined ) {

  if ( arr === undefined ) return arr;
  else if (arr.length === 0) return undefined;
  return arr;
}