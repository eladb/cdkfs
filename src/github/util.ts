export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

/**
 * Bash if equivalent execution of scripts that is compatible with shx.
 * Execution will fail, if the when or then command fails.
 *
 * @param when The condition to evaluate. Usually a `test` command. This command may be executed twice. Make sure it is non mutating.
 * @param then The script to run if the condition is true.
 * @param otherwise The script to run if the condition is false.
 */
export function shxIf(when: string, then: string, otherwise?: string): string {
  // Connect the condition and the 'then' command with a logical AND.
  // If the condition is true, this will execute the 'then' command.
  // This case is equivalent to TRUE && then()
  // If the condition is false, this will exit with a non-zero status code
  // (indicating an error) without executing the 'then' command.
  // This case is equivalent to FALSE && then()
  const if_clause = `${when} && (${then})`;

  // For the else clause, we need to check the 'when' condition again,
  // because the if clause can also return FALSE if the 'then' command fails.
  // However in this situation we don't want to execute the else clause but report an error.
  const else_clause = `${when} || (${otherwise})`;

  // Try the if_clause first. If it fails try the else clause.
  if (otherwise) {
    return `(${if_clause}) || (${else_clause})`;
  }

  // Just run the if clause
  return if_clause;
}
