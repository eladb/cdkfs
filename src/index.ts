export * from './file';
export * from './github-workflow';
export * from './ignore-file';
export * from './jsii-project';
export * from './json';
export * from './license';
export * from './node-project';
export * from './project';
export * from './component';
export * from './semver';
export * from './version';
export * from './eslint';
export * from './jest';
export * from './typescript';
export * from './mergify';
export * from './dependabot';