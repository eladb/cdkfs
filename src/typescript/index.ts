export * from './projenrc';
export * from './typescript';
export * from './typescript-config';
export * from './typescript-typedoc';
