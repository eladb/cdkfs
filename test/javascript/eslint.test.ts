import { execSync } from "child_process";
import { readFileSync } from "fs";
import * as path from "path";
import { SampleFile, TaskRuntime } from "../../src";
import {
  Eslint,
  EslintConfigFileFormat,
  NodeProject,
} from "../../src/javascript";
import { TypeScriptProject } from "../../src/typescript";
import { synthSnapshot, withProjectDir } from "../util";

test("devdirs", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "master",
  });

  // WHEN
  new Eslint(project, {
    devdirs: ["foo", "bar"],
    dirs: ["mysrc"],
    lintProjenRc: false,
  });

  // THEN
  expect(synthSnapshot(project)[".eslintrc.json"]).toMatchSnapshot();
});

describe("prettier", () => {
  test("snapshot", () => {
    // GIVEN
    const project = new NodeProject({
      name: "testProject",
      defaultReleaseBranch: "master",
    });

    // WHEN
    new Eslint(project, {
      dirs: ["mysrc"],
      prettier: true,
      lintProjenRc: false,
    });

    // THEN
    expect(synthSnapshot(project)[".eslintrc.json"]).toMatchSnapshot();
  });
});

describe("alias", () => {
  test("custom config", () => {
    // GIVEN
    const project = new NodeProject({
      name: "testProject",
      defaultReleaseBranch: "master",
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["mysrc"],
      aliasMap: {
        "@src": "./src",
        "@foo": "./src/foo",
      },
      aliasExtensions: [".ts", ".js"],
      lintProjenRc: false,
    });

    // THEN
    expect(eslint.config.settings["import/resolver"].alias).toHaveProperty(
      "map",
      [
        ["@src", "./src"],
        ["@foo", "./src/foo"],
      ]
    );
    expect(eslint.config.settings["import/resolver"].alias).toHaveProperty(
      "extensions",
      [".ts", ".js"]
    );
  });
});

test("tsAlwaysTryTypes", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "master",
  });

  // WHEN
  const eslint = new Eslint(project, {
    dirs: ["mysrc"],
    tsAlwaysTryTypes: true,
    lintProjenRc: false,
  });

  // THEN
  expect(eslint.config.settings["import/resolver"].typescript).toHaveProperty(
    "alwaysTryTypes",
    true
  );
});

test("if the prettier is configured, eslint is configured accordingly", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "master",
    prettier: true,
  });

  // WHEN
  new Eslint(project, {
    dirs: ["src"],
    lintProjenRc: false,
  });

  // THEN
  const output = synthSnapshot(project);
  expect(output[".eslintrc.json"].extends).toContain(
    "plugin:prettier/recommended"
  );
});

test("can output yml instead of json", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "main",
    prettier: true,
  });

  // WHEN
  new Eslint(project, {
    dirs: ["src"],
    yaml: true,
    lintProjenRc: false,
  });

  // THEN
  const output = synthSnapshot(project);
  expect(output[".eslintrc.yml"]).toBeDefined();
  expect(output[".eslintrc.json"]).toBeUndefined();
});

test("can override the parser", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "master",
    prettier: true,
  });

  // WHEN
  const eslint = new Eslint(project, {
    dirs: ["src"],
    lintProjenRc: false,
  });
  eslint.addOverride({
    files: ["*.json", "*.json5", "*.jsonc"],
    parser: "jsonc-eslint-parser",
  });
  const output = synthSnapshot(project);

  // THEN
  expect(output[".eslintrc.json"].overrides).toContainEqual({
    files: ["*.json", "*.json5", "*.jsonc"],
    parser: "jsonc-eslint-parser",
  });
});

test("creates a eslint task", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "master",
    prettier: true,
  });

  // WHEN
  const eslint = new Eslint(project, {
    dirs: ["src"],
    lintProjenRc: false,
  });

  // THEN
  const manifest = synthSnapshot(project)[TaskRuntime.MANIFEST_FILE];
  expect(eslint.eslintTask._renderSpec()).toMatchObject(manifest.tasks.eslint);
});

test("omit --ext when no extensions are specified", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "master",
    prettier: true,
  });

  // WHEN
  const eslint = new Eslint(project, {
    dirs: ["src"],
    lintProjenRc: false,
    fileExtensions: [],
  });

  // THEN
  const taskStep = eslint.eslintTask.steps[0];
  expect(taskStep.exec).not.toContain("--ext");
  expect(taskStep?.args ?? []).not.toContain(expect.stringContaining("--ext"));
});

test("add --ext when extensions are specified", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "master",
    prettier: true,
  });

  // WHEN
  const eslint = new Eslint(project, {
    dirs: ["src"],
    lintProjenRc: false,
  });

  // THEN
  const taskStep = eslint.eslintTask.steps[0];
  expect(taskStep.exec).toContain("--ext");
});

test("allow modification of the eslint task", () => {
  // GIVEN
  const project = new NodeProject({
    name: "testProject",
    defaultReleaseBranch: "master",
    prettier: true,
  });

  // WHEN
  const eslint = new Eslint(project, {
    dirs: ["src"],
    lintProjenRc: false,
  });

  const taskStep = eslint.eslintTask.steps[0];
  const newTestArg = "--foo";
  eslint.eslintTask.reset(taskStep.exec, { args: [newTestArg] });

  eslint.addLintPattern("bar");

  // THEN
  expect(eslint.eslintTask.steps[0].args).toContain(newTestArg);
});

test.each([
  EslintConfigFileFormat.JSON,
  EslintConfigFileFormat.JAVASCRIPT_OLD_CJS,
  EslintConfigFileFormat.JAVASCRIPT_FLAT_CJS,
  EslintConfigFileFormat.JAVASCRIPT_FLAT_ESM,
])(
  "eslint snapshot matches format '%s'",
  (fileFormat: EslintConfigFileFormat) => {
    // GIVEN
    const project = new NodeProject({
      name: "testProject",
      defaultReleaseBranch: "master",
      prettier: true,
    });

    // WHEN
    const eslint = new Eslint(project, {
      dirs: ["src"],
      lintProjenRc: false,
      fileFormat,
    });
    eslint.addExtends("./node_modules/coding-standard/eslintDefaults.js");
    eslint.addExtends("./node_modules/coding-standard/.eslintrc-es6");
    eslint.addExtends("./projen/eslintrc.js");

    const output = synthSnapshot(project);

    // THEN
    expect(output[eslint.fileName]).toMatchSnapshot();
  }
);

test.each([
  EslintConfigFileFormat.JSON,
  EslintConfigFileFormat.YAML,
  EslintConfigFileFormat.JAVASCRIPT_OLD_CJS,
  EslintConfigFileFormat.JAVASCRIPT_FLAT_CJS,
  EslintConfigFileFormat.JAVASCRIPT_FLAT_ESM,
])("eslint runs with %s", (fileFormat: EslintConfigFileFormat) => {
  withProjectDir((projectdir) => {
    const project = new TypeScriptProject({
      defaultReleaseBranch: "main",
      name: "testProject",
      outdir: projectdir,
      eslint: true,
      eslintOptions: {
        fileFormat,
      },
      release: true,
    });
    const testFilename = "src/format-test.ts";
    const contents = "    export  const  foo  =  { 'bar'  :   \"baz\",\n\n }  ";
    new SampleFile(project, testFilename, {
      contents,
    });
    project.synth();

    const fullTestFilename = path.join(projectdir, testFilename);
    expect(readFileSync(fullTestFilename).toString()).toEqual(contents);
    execSync("npx projen eslint", { cwd: projectdir, stdio: "inherit" });
    expect(readFileSync(fullTestFilename).toString()).toEqual(
      "export const foo = { bar: 'baz' };"
    );
  });
});
