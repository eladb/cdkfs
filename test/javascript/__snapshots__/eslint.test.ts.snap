// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`devdirs 1`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/typescript",
  ],
  "ignorePatterns": [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "./coverage",
    "!.projenrc.js",
  ],
  "overrides": [
    {
      "files": [
        ".projenrc.js",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
    "import",
  ],
  "root": true,
  "rules": {
    "@typescript-eslint/indent": [
      "error",
      2,
    ],
    "@typescript-eslint/member-delimiter-style": [
      "error",
    ],
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": [
      "error",
    ],
    "@typescript-eslint/no-require-imports": [
      "error",
    ],
    "@typescript-eslint/no-shadow": [
      "error",
    ],
    "@typescript-eslint/return-await": [
      "error",
    ],
    "array-bracket-newline": [
      "error",
      "consistent",
    ],
    "array-bracket-spacing": [
      "error",
      "never",
    ],
    "brace-style": [
      "error",
      "1tbs",
      {
        "allowSingleLine": true,
      },
    ],
    "comma-dangle": [
      "error",
      "always-multiline",
    ],
    "comma-spacing": [
      "error",
      {
        "after": true,
        "before": false,
      },
    ],
    "curly": [
      "error",
      "multi-line",
      "consistent",
    ],
    "dot-notation": [
      "error",
    ],
    "import/no-duplicates": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/foo/**",
          "**/bar/**",
        ],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "indent": [
      "off",
    ],
    "key-spacing": [
      "error",
    ],
    "keyword-spacing": [
      "error",
    ],
    "max-len": [
      "error",
      {
        "code": 150,
        "ignoreComments": true,
        "ignoreRegExpLiterals": true,
        "ignoreStrings": true,
        "ignoreTemplateLiterals": true,
        "ignoreUrls": true,
      },
    ],
    "no-bitwise": [
      "error",
    ],
    "no-multi-spaces": [
      "error",
      {
        "ignoreEOLComments": false,
      },
    ],
    "no-multiple-empty-lines": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
    "no-trailing-spaces": [
      "error",
    ],
    "object-curly-newline": [
      "error",
      {
        "consistent": true,
        "multiline": true,
      },
    ],
    "object-curly-spacing": [
      "error",
      "always",
    ],
    "object-property-newline": [
      "error",
      {
        "allowAllPropertiesOnSameLine": true,
      },
    ],
    "quote-props": [
      "error",
      "consistent-as-needed",
    ],
    "quotes": [
      "error",
      "single",
      {
        "avoidEscape": true,
      },
    ],
    "semi": [
      "error",
      "always",
    ],
    "space-before-blocks": [
      "error",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "node": {},
      "typescript": {
        "alwaysTryTypes": true,
        "project": "./tsconfig.json",
      },
    },
  },
}
`;

exports[`eslint snapshot matches format 'flat-cjs' 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
const pluginImport = require('eslint-plugin-import');
const pluginPrettier = require('eslint-plugin-prettier');
const extendsCodingStandardEslintDefaults = require('./node_modules/coding-standard/eslintDefaults.js');
const extendsCodingStandardEslintrcEs6 = require('./node_modules/coding-standard/.eslintrc-es6');
const extendsProjenEslintrc = require('./projen/eslintrc.js');
const pluginTypescriptEslint = require('@typescript-eslint/eslint-plugin');
const parserTypescriptEslintParser = require('@typescript-eslint/parser');
module.exports = [
  {
    files: [
      "**/*.ts",
    ],
    ...pluginImport.configs.typescript,
  },
  {
    files: [
      "**/*.ts",
    ],
    ...pluginPrettier.configs.recommended,
  },
  {
    files: [
      "**/*.ts",
    ],
    ...extendsCodingStandardEslintDefaults,
  },
  {
    files: [
      "**/*.ts",
    ],
    ...extendsCodingStandardEslintrcEs6,
  },
  {
    files: [
      "**/*.ts",
    ],
    ...extendsProjenEslintrc,
  },
  {
    plugins: {
      "@typescript-eslint": pluginTypescriptEslint,
      import: pluginImport,
    },
    settings: {
      "import/parsers": {
        "@typescript-eslint/parser": [
          ".ts",
          ".tsx",
        ],
      },
      "import/resolver": {
        node: {},
        typescript: {
          project: "./tsconfig.json",
          alwaysTryTypes: true,
        },
      },
    },
    rules: {
      "@typescript-eslint/no-require-imports": [
        "error",
      ],
      "import/no-extraneous-dependencies": [
        "error",
        {
          devDependencies: [],
          optionalDependencies: false,
          peerDependencies: true,
        },
      ],
      "import/no-unresolved": [
        "error",
      ],
      "import/order": [
        "warn",
        {
          groups: [
            "builtin",
            "external",
          ],
          alphabetize: {
            order: "asc",
            caseInsensitive: true,
          },
        },
      ],
      "no-duplicate-imports": [
        "error",
      ],
      "no-shadow": [
        "off",
      ],
      "@typescript-eslint/no-shadow": [
        "error",
      ],
      "key-spacing": [
        "error",
      ],
      "no-multiple-empty-lines": [
        "error",
      ],
      "@typescript-eslint/no-floating-promises": [
        "error",
      ],
      "no-return-await": [
        "off",
      ],
      "@typescript-eslint/return-await": [
        "error",
      ],
      "no-trailing-spaces": [
        "error",
      ],
      "dot-notation": [
        "error",
      ],
      "no-bitwise": [
        "error",
      ],
      "@typescript-eslint/member-ordering": [
        "error",
        {
          default: [
            "public-static-field",
            "public-static-method",
            "protected-static-field",
            "protected-static-method",
            "private-static-field",
            "private-static-method",
            "field",
            "constructor",
            "method",
          ],
        },
      ],
    },
    files: [
      "**/*.ts",
    ],
    ignores: [
      "**/*.js",
      "**/*.d.ts",
      "node_modules/",
      "**/*.generated.ts",
      "./coverage",
      "!.projenrc.js",
    ],
    languageOptions: {
      parser: parserTypescriptEslintParser,
      parserOptions: {
        ecmaVersion: 2018,
        sourceType: "module",
        project: "./tsconfig.json",
      },
      globals: {
        jest: true,
        node: true,
      },
    },
  },
  {
    files: [
      ".projenrc.js",
    ],
    rules: {
      "@typescript-eslint/no-require-imports": "off",
      "import/no-extraneous-dependencies": "off",
    },
  },
];
"
`;

exports[`eslint snapshot matches format 'flat-esm' 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import pluginImport from 'eslint-plugin-import';
import pluginPrettier from 'eslint-plugin-prettier';
import extendsCodingStandardEslintDefaults from './node_modules/coding-standard/eslintDefaults.js';
import extendsCodingStandardEslintrcEs6 from './node_modules/coding-standard/.eslintrc-es6';
import extendsProjenEslintrc from './projen/eslintrc.js';
import pluginTypescriptEslint from '@typescript-eslint/eslint-plugin';
import parserTypescriptEslintParser from '@typescript-eslint/parser';
export default [
  {
    files: [
      "**/*.ts",
    ],
    ...pluginImport.configs.typescript,
  },
  {
    files: [
      "**/*.ts",
    ],
    ...pluginPrettier.configs.recommended,
  },
  {
    files: [
      "**/*.ts",
    ],
    ...extendsCodingStandardEslintDefaults,
  },
  {
    files: [
      "**/*.ts",
    ],
    ...extendsCodingStandardEslintrcEs6,
  },
  {
    files: [
      "**/*.ts",
    ],
    ...extendsProjenEslintrc,
  },
  {
    plugins: {
      "@typescript-eslint": pluginTypescriptEslint,
      import: pluginImport,
    },
    settings: {
      "import/parsers": {
        "@typescript-eslint/parser": [
          ".ts",
          ".tsx",
        ],
      },
      "import/resolver": {
        node: {},
        typescript: {
          project: "./tsconfig.json",
          alwaysTryTypes: true,
        },
      },
    },
    rules: {
      "@typescript-eslint/no-require-imports": [
        "error",
      ],
      "import/no-extraneous-dependencies": [
        "error",
        {
          devDependencies: [],
          optionalDependencies: false,
          peerDependencies: true,
        },
      ],
      "import/no-unresolved": [
        "error",
      ],
      "import/order": [
        "warn",
        {
          groups: [
            "builtin",
            "external",
          ],
          alphabetize: {
            order: "asc",
            caseInsensitive: true,
          },
        },
      ],
      "no-duplicate-imports": [
        "error",
      ],
      "no-shadow": [
        "off",
      ],
      "@typescript-eslint/no-shadow": [
        "error",
      ],
      "key-spacing": [
        "error",
      ],
      "no-multiple-empty-lines": [
        "error",
      ],
      "@typescript-eslint/no-floating-promises": [
        "error",
      ],
      "no-return-await": [
        "off",
      ],
      "@typescript-eslint/return-await": [
        "error",
      ],
      "no-trailing-spaces": [
        "error",
      ],
      "dot-notation": [
        "error",
      ],
      "no-bitwise": [
        "error",
      ],
      "@typescript-eslint/member-ordering": [
        "error",
        {
          default: [
            "public-static-field",
            "public-static-method",
            "protected-static-field",
            "protected-static-method",
            "private-static-field",
            "private-static-method",
            "field",
            "constructor",
            "method",
          ],
        },
      ],
    },
    files: [
      "**/*.ts",
    ],
    ignores: [
      "**/*.js",
      "**/*.d.ts",
      "node_modules/",
      "**/*.generated.ts",
      "./coverage",
      "!.projenrc.js",
    ],
    languageOptions: {
      parser: parserTypescriptEslintParser,
      parserOptions: {
        ecmaVersion: 2018,
        sourceType: "module",
        project: "./tsconfig.json",
      },
      globals: {
        jest: true,
        node: true,
      },
    },
  },
  {
    files: [
      ".projenrc.js",
    ],
    rules: {
      "@typescript-eslint/no-require-imports": "off",
      "import/no-extraneous-dependencies": "off",
    },
  },
];
"
`;

exports[`eslint snapshot matches format 'json' 1`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/typescript",
    "plugin:prettier/recommended",
    "./node_modules/coding-standard/eslintDefaults.js",
    "./node_modules/coding-standard/.eslintrc-es6",
    "./projen/eslintrc.js",
  ],
  "ignorePatterns": [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "./coverage",
    "!.projenrc.js",
  ],
  "overrides": [
    {
      "files": [
        ".projenrc.js",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
    "import",
  ],
  "root": true,
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": [
      "error",
    ],
    "@typescript-eslint/no-require-imports": [
      "error",
    ],
    "@typescript-eslint/no-shadow": [
      "error",
    ],
    "@typescript-eslint/return-await": [
      "error",
    ],
    "dot-notation": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "key-spacing": [
      "error",
    ],
    "no-bitwise": [
      "error",
    ],
    "no-duplicate-imports": [
      "error",
    ],
    "no-multiple-empty-lines": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
    "no-trailing-spaces": [
      "error",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "node": {},
      "typescript": {
        "alwaysTryTypes": true,
        "project": "./tsconfig.json",
      },
    },
  },
}
`;

exports[`eslint snapshot matches format 'old-cjs' 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

module.exports = {
  env: {
    jest: true,
    node: true,
  },
  root: true,
  plugins: [
    "@typescript-eslint",
    "import",
  ],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: "module",
    project: "./tsconfig.json",
  },
  extends: [
    "plugin:import/typescript",
    "plugin:prettier/recommended",
    "./node_modules/coding-standard/eslintDefaults.js",
    "./node_modules/coding-standard/.eslintrc-es6",
    "./projen/eslintrc.js",
  ],
  settings: {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      node: {},
      typescript: {
        project: "./tsconfig.json",
        alwaysTryTypes: true,
      },
    },
  },
  ignorePatterns: [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "./coverage",
    "!.projenrc.js",
  ],
  rules: {
    "@typescript-eslint/no-require-imports": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        devDependencies: [],
        optionalDependencies: false,
        peerDependencies: true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        groups: [
          "builtin",
          "external",
        ],
        alphabetize: {
          order: "asc",
          caseInsensitive: true,
        },
      },
    ],
    "no-duplicate-imports": [
      "error",
    ],
    "no-shadow": [
      "off",
    ],
    "@typescript-eslint/no-shadow": [
      "error",
    ],
    "key-spacing": [
      "error",
    ],
    "no-multiple-empty-lines": [
      "error",
    ],
    "@typescript-eslint/no-floating-promises": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "@typescript-eslint/return-await": [
      "error",
    ],
    "no-trailing-spaces": [
      "error",
    ],
    "dot-notation": [
      "error",
    ],
    "no-bitwise": [
      "error",
    ],
    "@typescript-eslint/member-ordering": [
      "error",
      {
        default: [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
  },
  overrides: [
    {
      files: [
        ".projenrc.js",
      ],
      rules: {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
};
"
`;

exports[`prettier snapshot 1`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/typescript",
    "plugin:prettier/recommended",
  ],
  "ignorePatterns": [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "./coverage",
    "!.projenrc.js",
  ],
  "overrides": [
    {
      "files": [
        ".projenrc.js",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
    "import",
  ],
  "root": true,
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": [
      "error",
    ],
    "@typescript-eslint/no-require-imports": [
      "error",
    ],
    "@typescript-eslint/no-shadow": [
      "error",
    ],
    "@typescript-eslint/return-await": [
      "error",
    ],
    "dot-notation": [
      "error",
    ],
    "import/no-duplicates": [
      "error",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "key-spacing": [
      "error",
    ],
    "no-bitwise": [
      "error",
    ],
    "no-multiple-empty-lines": [
      "error",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
    "no-trailing-spaces": [
      "error",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "node": {},
      "typescript": {
        "alwaysTryTypes": true,
        "project": "./tsconfig.json",
      },
    },
  },
}
`;
